### PR TITLE
[SYCL][COMPAT] Added filter_device and list_devices

### PR DESCRIPTION
**Pull request by Alcpz**
This PR adds functionalities for:
* Listing devices in stdout
* Filtering devices

Tests and docs updated accordingly.

### CODE REVIEW

File: sycl/include/syclcompat/device.hpp [hunk 1]
```diff
@@ -662,14 +662,64 @@ class dev_mgr {
   unsigned int device_count() { return _devs.size(); }
 
   unsigned int get_device_id(const sycl::device &dev) {
+    if (!_devs.size()) {
+      throw std::runtime_error(
+          "[SYCLcompat] No SYCL devices found in the device list.");
```
Comment (joeatodd) [inline sycl/include/syclcompat/device.hpp:667]:
```suggestion
          "[SYCLcompat] No SYCL devices found in the device list. Device list may have been filtered by syclcompat::filter_device");
```

Comment (joeatodd) [reply sycl/include/syclcompat/device.hpp:667]: Or something like that. Effectively this is the same error as the `runtime_error` below, just with fewer (none) other devices.

File: sycl/test-e2e/syclcompat/device/device_fixt.hpp [hunk 2]
```diff
@@ -50,3 +50,32 @@ class DeviceExtFixt {
 
   syclcompat::device_ext &get_dev_ext() { return dev_; }
 };
+
+// Helper for counting the output lines of syclcompat::list_devices
+// Used to override std::cout
+class CountingStream : public std::streambuf {
```
Comment (joeatodd) [inline sycl/test-e2e/syclcompat/device/device_fixt.hpp:56]: This is cool!

File: sycl/test-e2e/syclcompat/device/device_filter.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// RUN: %{run} %t.out
+
+#include <syclcompat/device.hpp>
+
+void test_filter_devices() {
+  auto &dev = syclcompat::get_current_device();
+
+  assert(syclcompat::detail::dev_mgr::instance().device_count() > 0);
+
+  syclcompat::filter_device({"NON-EXISTENT DEVICE"});
```
Comment (joeatodd) [inline sycl/test-e2e/syclcompat/device/device_filter.cpp:33]: We could also test to ensure that `syclcompat::filter_device(dev.get_info<sycl::info::device::name>())` doesn't filter the current device?

File: sycl/doc/syclcompat/README.md [hunk 4]
```diff
@@ -831,6 +831,15 @@ static inline sycl::context get_default_context();
 // Util function to get a CPU device.
 static inline device_ext &cpu_device();
 
+/// Filter out devices; only keep the device whose name contains one of the
+/// subname in \p dev_subnames.
+/// May break device id mapping and change current device. It's better to be
+/// called before other SYCLcompat or SYCL APIs.
+static inline void filter_device(const std::vector<std::string> &dev_subnames);
+
+/// List all the devices with its id in dev_mgr.
```
Comment (joeatodd) [inline sycl/doc/syclcompat/README.md:840]:
```suggestion
/// Print all the devices (and their IDs) in the dev_mgr
```

File: sycl/doc/syclcompat/README.md [hunk 5]
```diff
@@ -855,6 +864,12 @@ can be queried through `device_ext` as well.
 throws a `sycl::exception` if the device does not have the specified list of
 `sycl::aspect`.
 
+Devices can be listed and filtered using `syclcompat::list_devices()` and
+`syclcompat::filter_device()`. If `SYCLCOMPAT_VERBOSE` is defined at compile
+time, the available SYCL devices are printed to the standard output both at
+initialization time, or if the device list is filtered using
```
Comment (joeatodd) [inline sycl/doc/syclcompat/README.md:870]:
```suggestion
initialization time, and when the device list is filtered using
```

**Review by joeatodd — CHANGES_REQUESTED** [review]
An extra test and a couple of language things, otherwise LGTM :+1:

File: sycl/test-e2e/syclcompat/device/device_filter.cpp (hunk 3 above)
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/device/device_filter.cpp:44]: Added

File: sycl/test-e2e/syclcompat/device/device_filter.cpp [hunk 6]
```diff
@@ -25,6 +25,17 @@
...
 
+void test_filtering_existing_device() {
+  auto &dev = syclcompat::get_current_device();
+  std::string dev_name = dev.get_info<sycl::info::device::name>();
+  std::string dev_substr = dev_name.substr(0, dev_name.find(" ") + 2);
+  int n_devices = syclcompat::detail::dev_mgr::instance().device_count();
+
+  syclcompat::filter_device({dev_substr});
+
+  assert(syclcompat::detail::dev_mgr::instance().device_count() == n_devices);
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/device/device_filter.cpp]: Updated. Also checks for the full name of the device apart from a substring.

**Review by joeatodd — APPROVED** [review]

**Comment by Alcpz** [comment]
@intel/llvm-gatekeepers Ready for merge. Thanks!

**Comment by Alcpz** [comment]
@intel/llvm-gatekeepers this is ready to merge. Thanks!

**Pull request merged**